controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+	return labelValue == util.MustFormatValue(rsName)
+}
 
-		// Handle deletion of replicas
+// getNewVirtualMachine creates a new VirtualMachine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func (r *Reconciler) getNewVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet) *vmopv1.VirtualMachine {
+	// gv := vmopv1.SchemeGroupVersion
+	vm := &vmopv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", rs.Name),
+			// Note: by setting the ownerRef on creation we signal to the
+			// VirtualMachine controller that this is not a stand-alone VM.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(rs, replicaSetKind)},
+			Namespace:       rs.Namespace,
+			Labels: map[string]string{
+				vmopv1.VirtualMachineReplicaSetNameLabel: util.MustFormatValue(rs.Name),
+			},
+			Annotations: rs.Spec.Template.Annotations,
+		},
+		// TypeMeta: metav1.TypeMeta{
+		// 	Kind:       gv.WithKind("VirtualMachine").Kind,
+		// 	APIVersion: gv.String(),
+		// },
+		Spec: rs.Spec.Template.Spec,
+	}
 
-		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	vm.Labels = getLabelsFromVMReplicaSet(rs)
+
+	vm.Annotations = getAnnotationsFromReplicaSet(rs)
+
+	// TODO: Propagate the VirtualMachineDeploymentNameLabel from VirtualMachineReplicaSet to VirtualMachines if it exists.
+
+	return vm
+}
+
+func getLabelsFromVMReplicaSet(rs *vmopv1.VirtualMachineReplicaSet) map[string]string {
+	labels := make(map[string]string)
+
+	// Copy over the labels from the template
+	maps.Copy(labels, rs.Spec.Template.Labels)
+
+	// Ensure that the ReplicaSetLabel is always present.
+	labels[vmopv1.VirtualMachineReplicaSetNameLabel] = util.MustFormatValue(rs.Name)
+
+	return labels
+}
+
+func getAnnotationsFromReplicaSet(rs *vmopv1.VirtualMachineReplicaSet) map[string]string {
+	annotations := make(map[string]string)
+
+	// Copy over the labels from the template
+	maps.Copy(annotations, rs.Spec.Template.Annotations)
+
+	return annotations
+}
+
+// syncReplicas scales VirtualMachine resources up or down.
+func (r *Reconciler) syncReplicas(
+	ctx *pkgctx.VirtualMachineReplicaSetContext,
+	rs *vmopv1.VirtualMachineReplicaSet,
+	vms []*vmopv1.VirtualMachine) error {
+
+	if rs.Spec.Replicas == nil {
+		return errors.Errorf("the Replicas field in Spec for VirtualMachineReplicaSet %v is nil, this should not be allowed", rs.Name)
+	}
+	diff := len(vms) - int(*(rs.Spec.Replicas))
+	switch {
+	case diff < 0:
+		diff *= -1
+		ctx.Logger.Info("ReplicaSet is scaling up",
+			"currentReplicas", len(vms),
+			"desiredReplicas", *rs.Spec.Replicas,
+			"vmsToBeCreated", diff,
+		)
+
+		var (
+			vmList []*vmopv1.VirtualMachine
+			errs   []error
+		)
+
+		for i := 0; i < diff; i++ {
+			vm := r.getNewVirtualMachine(rs)
+			log := ctx.Logger.WithValues("vm", vm.Name)
+
+			log.Info(fmt.Sprintf("Creating VM %d of %d", i+1, diff))
```

Comment:
Also, given that it is logged below as well as `Created VM`, how about we pick one of these statements and log it at a lower log level so we do not spam the logs? I usually prefer to log attempt statements at the higher log level and completions at lower log levels. Because generally, if we do not see an error, we know it completed. And if we are trying to gather timings, we can turn up logging.